drivers/adc/adc_gd32.c

```diff
@@ -0,0 +1,435 @@
+/*
+ * Copyright (c) 2022 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_adc
+
+#include <errno.h>
+#include <drivers/pinctrl.h>
+#include <drivers/adc.h>
+#include <zephyr/devicetree.h>
+#include <soc.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_gd32, CONFIG_ADC_LOG_LEVEL);
+
+#ifndef	ADC0
+/**
+ * @brief The name of gd32 ADC HAL are different between single and multi ADC SoCs.
+ * This adjust the single ADC SoC HAL, so we can call gd32 ADC HAL in a common way.
+ */
+#undef ADC_STAT
+#undef ADC_CTL0
+#undef ADC_CTL1
+#undef ADC_SAMPT0
+#undef ADC_SAMPT1
+#undef ADC_RSQ2
+#undef ADC_RDATA
+
+#define ADC0			ADC
+#define ADC_STAT(adc0)		REG32((adc0) + 0x00000000U)
+#define ADC_CTL0(adc0)		REG32((adc0) + 0x00000004U)
+#define ADC_CTL1(adc0)		REG32((adc0) + 0x00000008U)
+#define ADC_SAMPT0(adc0)	REG32((adc0) + 0x0000000CU)
+#define ADC_SAMPT1(adc0)	REG32((adc0) + 0x00000010U)
+#define ADC_RSQ2(adc0)		REG32((adc0) + 0x00000034U)
+#define ADC_RDATA(adc0)		REG32((adc0) + 0x0000004CU)
+#endif
+
+#define SPT_WIDTH	3U
+#define SAMPT1_SIZE	10U
+
+#if defined(CONFIG_SOC_SERIES_GD32F4XX)
+#define SMP_TIME(x)	ADC_SAMPLETIME_##x
+
+static const uint16_t acq_time_tbl[8] = {3, 15, 28, 56, 84, 112, 144, 480};
+static const uint32_t table_samp_time[] = {
+	SMP_TIME(3),
+	SMP_TIME(15),
+	SMP_TIME(28),
+	SMP_TIME(56),
+	SMP_TIME(84),
+	SMP_TIME(112),
+	SMP_TIME(144),
+	SMP_TIME(480)
+};
+#else
+#define SMP_TIME(x)	ADC_SAMPLETIME_##x##POINT5
+
+static const uint16_t acq_time_tbl[8] = {2, 8, 14, 29, 42, 56, 72, 240};
+static const uint32_t table_samp_time[] = {
+	SMP_TIME(1),
+	SMP_TIME(7),
+	SMP_TIME(13),
+	SMP_TIME(28),
+	SMP_TIME(41),
+	SMP_TIME(55),
+	SMP_TIME(71),
+	SMP_TIME(239)
+};
+#endif
+
+static uint8_t adc_gd32_init_irqs[8U];
```

Comment:
note: why is this array size set to 8? can this be explained?

Reply:
Use `GD32_ADC_NUM_MAX` macro instread. GD32 ADC number always smaller than 3.